easybuild/tools/toolchain/compiler.py

```diff
@@ -278,18 +282,43 @@ def _set_optimal_architecture(self, default_optarch=None):
                                 (--optarch and --optarch=GENERIC still override this value)
         """
         optarch = None
+        use_generic = False
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if build_option('optarch') is not None:
+            # optarch has been validated as a simple string
+            if type(build_option('optarch')) is str:
+                # Option is generic, so mark it as to be set later
+                if build_option('optarch') == OPTARCH_GENERIC:
+                    use_generic = True
+                # No compilers specified, and option is not generic, so simply take what is there
+                else:
+                    optarch = build_option('optarch')
+
+            # optarch has been validated as complex string with multiple compilers and converted to a dictionary
+            elif type(build_option('optarch')) is dict:
+                current_compiler = getattr(self, 'COMPILER_FAMILY', None)
+                if current_compiler in build_option('optarch'):
+                    if build_option('optarch')[current_compiler] == OPTARCH_GENERIC:
+                        use_generic = True
+                    else:
+                        optarch = build_option('optarch')[current_compiler]
+                # no option for this compiler
+                else:
+                    pass
```

Comment:
this `pass` should be replaced with the `no optarch found` log message?

together with resetting `optarch` to `None`?